File: src/Y2025/Day10/main.py
from functools import reduce
from itertools import product
from operator import xor
from typing import TextIO

from attrs import frozen
from rich.pretty import pprint
from z3 import Int, Optimize, sat

@frozen
class Machine:
    config: int
    buttons: list[set[int]]
    joltages: list[int]

    @classmethod
    def parse(cls, line: str):
        config, *buttons, joltages = line.split(" ")
        return cls(
            sum(1 << i for i, v in enumerate(config[1:-1]) if v == "#"),
            [{*map(int, button[1:-1].split(","))} for button in buttons],
            [*map(int, joltages[1:-1].split(","))],
        )


def parser(raw_data: TextIO) -> list[Machine]:
    return [Machine.parse(line) for line in raw_data.read().splitlines()]


def full_solver(machines: list[Machine]) -> tuple[int, int]:
    p1 = p2 = 0
    for machine in machines:
        button_bits = [sum(1 << i for i in button) for button in machine.buttons]
        p1 += min(
            sum(indices)
            for indices in product(range(2), repeat=len(button_bits))
            if not reduce(
                xor,
                (button_bits[i] for i, v in enumerate(indices) if v),
                initial=machine.config,
            )
        )
        s = Optimize()
        for idx, joltage in enumerate(machine.joltages):
            s.add(
                sum(
                    Int(f"x{i}")
                    for i, button in enumerate(machine.buttons)
                    if idx in button
                )
                == joltage
            )
        for i in range(len(machine.buttons)):
            s.add(Int(f"x{i}") >= 0)
        h = s.minimize(sum(Int(f"x{i}") for i in range(len(machine.buttons))))
        if s.check() == sat:
            p2 += s.lower(h).as_long()
        else:
            raise ValueError("joever")

    return (p1, p2)


if __name__ == "__main__":
    testing = False

    try:
        from Tools.Python.path_stuff import test_path
    except ModuleNotFoundError:
        path = "input.txt"
    else:
        path = test_path if testing else "input.txt"

    with open(path, "r") as file:
        data = parser(file)

    pprint(full_solver(data))

File: src/Y2025/Day10/main.rs
use std::fs;
extern crate z3;
use std::cmp::min;
use std::collections::HashMap;
use z3::{Optimize, ast::Int};

#[derive(Debug)]
struct Machine {
    config: u16,
    buttons: Vec<u16>,
    joltages: Vec<u16>,
}

fn parse_line(line: &str) -> Machine {
    let mut config = 0;
    let mut buttons = Vec::new();
    let mut joltages = Vec::new();
    for part in line.split(" ") {
        match part.chars().next() {
            Some('[') => {
                config = part[1..part.len() - 1]
                    .chars()
                    .enumerate()
                    .map(|(i, v)| if v == '#' { 1 << i } else { 0 })
                    .sum();
            }
            Some('(') => {
                buttons.push(
                    part[1..part.len() - 1]
                        .split(",")
                        .map(|i| 1 << i.parse::<u8>().unwrap())
                        .sum(),
                );
            }
            Some('{') => {
                joltages = part[1..part.len() - 1]
                    .split(",")
                    .map(|i| i.parse::<u16>().unwrap())
                    .collect();
            }
            _ => panic!("joever"),
        }
    }
    Machine {
        config,
        buttons,
        joltages,
    }
}

fn parse(filename: &str) -> Vec<Machine> {
    let contents = fs::read_to_string(filename).expect("Bro did not cook");
    contents.lines().map(parse_line).collect()
}

fn solve_p1(machine: &Machine) -> u32 {
    let mut combos = HashMap::<u16, u32>::new();
    combos.insert(machine.config, 0);
    for button in &machine.buttons {
        let new_combos: HashMap<u16, u32> =
            combos.iter().map(|(k, v)| (k ^ button, v + 1)).collect();
        for (k, v) in new_combos {
            combos.entry(k).and_modify(|w| *w = min(v, *w)).or_insert(v);
        }
    }
    *combos.get(&0).unwrap_or(&42069)
}

fn solve_p2(machine: &Machine) -> u64 {
    let solver = Optimize::new();
    let variables: Vec<Int> = (0..machine.buttons.len())
        .map(|i| Int::fresh_const(&i.to_string()))
        .collect();
    for (idx, joltage) in machine.joltages.iter().enumerate() {
        let constraint: Int = (0..machine.buttons.len())
            .filter(|i| (machine.buttons[*i] >> idx) & 1 == 1)
            .map(|i| variables[i].clone())
            .sum();
        solver.assert(&constraint.eq(*joltage));
    }
    for variable in &variables {
        solver.assert(&variable.ge(0));
    }
    let handle: Int = variables.iter().cloned().sum();
    solver.minimize(&handle);
    solver.check(&[]);
    let Some(model) = solver.get_model() else {
        panic!("cooked")
    };
    model.eval(&handle, true).unwrap().as_u64().unwrap()
}

fn solve(machines: Vec<Machine>) -> (u32, u64) {
    let mut p1 = 0;
    let mut p2 = 0;
    for machine in machines {
        p1 += solve_p1(&machine);
        p2 += solve_p2(&machine);
    }
    (p1, p2)
}

fn main() {
    let graph = parse("input.txt");
    let (p1, p2) = solve(graph);
    println!("{p1} {p2}")
}
